%\VignetteIndexEntry{Bioconductor LaTeX Style}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}

\documentclass{beamer}

\usetheme{Boadilla}
\setbeamertemplate{footline}
{
  \leavevmode%
  \hbox{%
  \begin{beamercolorbox}[wd=.5\paperwidth,ht=2.25ex,dp=1ex,right,rightskip=1em]{section in head/foot}%
    \usebeamerfont{subsection in head/foot}\hspace*{2ex}\insertshorttitle
  \end{beamercolorbox}%
  \begin{beamercolorbox}[wd=.4\paperwidth,ht=2.25ex,dp=1ex,left,leftskip=1em]{subsection in head/foot}%
    \usebeamerfont{section in head/foot}\insertsectionhead\hspace*{2ex}
  \end{beamercolorbox}%
  \begin{beamercolorbox}[wd=.1\paperwidth,ht=2.25ex,dp=1ex,right]{date in head/foot}%
    \insertframenumber{} / \inserttotalframenumber\hspace*{2ex}
  \end{beamercolorbox}}%
  \vskip0pt%
}

\newcommand{\software}[1]{\textsl{#1}}
\newcommand\R{\software{R}}
%
\newcommand\Bioconductor{\software{Bioconductor}}
\newcommand{\Rpackage}[1]{\textsl{#1}}
\newcommand\Biocpkg[1]{%
  {\href{http://bioconductor.org/packages/#1}%
    {\Rpackage{#1}}}}
\newcommand\CRANpkg[1]{%
  {\href{http://cran.fhcrc.org/web/packages/#1/index.html}%
    {\Rpackage{#1}}}}
\newcommand{\Rcode}[1]{\texttt{#1}}
\newcommand{\Rfunction}[1]{\Rcode{#1}}
\newcommand{\Robject}[1]{\Rcode{#1}}
\newcommand{\Rclass}[1]{\textit{#1}}

\title{The \Bioconductor{} Project: Current Status}
\author{Martin Morgan}
\institute[RPCI]{
  Roswell Park Cancer Institute\\
  Buffalo, NY, USA\\
  \href{mailto:martin.morgan@roswellpark.org}{martin.morgan@roswellpark.org}}
\date{4 November 2016}

\begin{document}

\maketitle

\section{Introduction}

\begin{frame}{\Bioconductor{}}
  \begin{columns}
  \column{.5\textwidth}
  \includegraphics[width=.8\textwidth]{bioconductor_logo_rgb.eps}
  \bigskip\par
  \includegraphics[width=.15\textwidth]{Rlogo.png}
  \hspace{1em}
  \includegraphics[width=.3\textwidth]{nid61380-a1-rpci-cmyk-black.eps}
  \bigskip\par
  \includegraphics[width=!, height=2em]{logo_nhgri.png}
  \bigskip\par
  \includegraphics[width=.8\textwidth]{nci-logo-full.png}
  \bigskip\par
  \href{https://bioconductor.org}{https://bioconductor.org}\par
  \href{https://support.bioconductor.org}{https://support.bioconductor.org}
  \column{.5\textwidth}
  Analysis and comprehension of high-throughput genomic data.
  \begin{itemize}
  \item Started 2002
  \item 1296 \R{} packages -- developed by `us' and user-contributed.
  \end{itemize}
  Well-used and respected.
  \begin{itemize}
  \item 43k unique IP downloads / month.
  \item 17,000 PubMedCentral citations.
  \end{itemize}
  \end{columns}
\end{frame}

\section{State of the project}

\begin{frame}{State of the project}
  \begin{itemize}
  \item Packages
  \item Users
  \item Web \& support sites
  \item Training \& meetings
  \item Release \& devel builders
  \item Funding
  \item Governance: (annual) Scientific Advisory Board; (monthly)
    Technical Advisory Board
  \end{itemize}
\end{frame}

\section{Recent developments}

\begin{frame}{Recent developments}
  \begin{itemize}
  \item New package reviews
  \item \Biocpkg{ExperimentHub} and \Biocpkg{AnnotationHub}
  \item Large data representation: \Biocpkg{HDF5Array}
  \item (Sneak peak) \software{Organism.dplyr}
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{\Biocpkg{HDF5Array}}
  \begin{semiverbatim}
library(HDF5Array)   # available in release & devel
n = 10000; m = 1000; # very large size
h5 = HDF5Array(matrix(rnorm(n * m), n))   
h5 + h5              # 'delayed' computation
library(SummarizedExperiment)
SummarizedExperiment(h5) # rich context
  \end{semiverbatim}
\end{frame}

\begin{frame}[fragile]{Sneak peak: \software{Organism.dplyr}}
  \begin{semiverbatim}
> library(Organism.dplyr) # not yet publicly available
> src = src_ucsc("Homo sapiens")  # any org.* + TxDb.*
using org.Hs.eg.db, TxDb.Hsapiens.UCSC.hg38.knownGene
> src
src:  sqlite 3.8.6 [/home/mtmorgan/organism_dplry.sqlite]
tbls: id, id_accession, id_go, id_go_all, id_omim_pm,
  id_protein, id_transcript, ranges_cds, ranges_exon,
  ranges_gene, ranges_tx
> tbl(src, 'id') %>% filter(symbol == 'BRCA1') %>% 
    select(ensembl, symbol, genename)
> exons(src, filter=list(symobl='BRCA1'))      # GRanges
> exons_tbl(src, filter=list(symbol='BRCA1'))  # tibble
  \end{semiverbatim}
\end{frame}

\section{Programming best practices}

\begin{frame}{Programming best practices}
  \begin{itemize}
  \item Reuse \& interoperability
  \item Correct, robust, efficient (vectorized) code;
    \Biocpkg{BiocParallel}
  \item Documentation: classic or \CRANpkg{roxygen2}
  \item Testing: \CRANpkg{RUnit} or \CRANpkg{testthat}
  \item Classic, tidy, and semantically rich data
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{Correct, robust, efficient\ldots}
  \begin{columns}[t]
    \column{.5\textwidth}
    \begin{verbatim}
f = function(n) {
  x = integer(0)
  for (i in 1:n)
      x = c(x, i)
  x
}
microbenchmark(f(1000),
  f(10000), f(100000))

f1 = function(n) {
  x = integer(n)
  for (i in 1:n)
    x[i] = i
  x
}
    \end{verbatim}
    \column{.5\textwidth}
    \begin{semiverbatim}
f2 = function(n)
  vapply(1:n, c, integer(1))

f3 = function(n)
  seq_len(n)

## correct
identical(f(100), f3(100))

## robust!
f(0); f3(0)                 

## efficient
system.time(f3(1e9)         
    \end{semiverbatim}
  \end{columns}

\end{frame}

\begin{frame}{Classic, tidy, rich: RNA-seq count data}
  Classic
  \begin{itemize}
  \item Sample x (phenotype + expression) Feature \Rcode{data.frame}
  \end{itemize}
  Tidy
  \begin{itemize}
  \item 'Melt' expression values to two long columns, replicated
    phenotype columns. End result: long data frame.
  \end{itemize}
  Rich, e.g., \Rcode{SummarizedExperiment}
  \begin{itemize}
  \item Phenotype and expression data manipulated in a coordinated
    fashion but stored separately.
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{Classic, tidy, rich: RNA-seq count data}
  \begin{semiverbatim}
df0 <- as.data.frame(list(mean=colMeans(classic[, -(1:22)])))
df1 <- tidy %>% group_by(probeset) %>% 
    summarize(mean=mean(exprs))
df2 <- as.data.frame(list(mean=rowMeans(assay(rich))))
ggplot(df1, aes(mean)) + geom_density()
  \end{semiverbatim}
\end{frame}

\begin{frame}{Classic, tidy, rich: RNA-seq count data}
  \begin{columns}[t]
    \column{.5\textwidth}
    Vocabulary
    \begin{itemize}
    \item Classic: extensive
    \item Tidy: restricted ‘endomorphisms’
    \item Rich: extensive, ‘meaningful’
    \end{itemize}
    Constraints (e.g., probes \& samples)
    \begin{itemize}
    \item Tidy: implicit
    \item Classic, Rich: explicit
    \end{itemize}
    Flexibility
    \begin{itemize}
    \item Classic, tidy: general-purpose
    \item Rich: specialized
    \end{itemize}

    \column{.5\textwidth}
    Programming contract
    \begin{itemize}
    \item   Classic, tidy: limited
    \item Rich: strict
    \end{itemize}
    Lessons learned / best practices
    \begin{itemize}
    \item Considerable value in semantically rich structures
    \item Current implementations trade-off user and developer
      convenience
    \item Endomorphism, simple vocabulary, consistent paradigm aid use
    \end{itemize}
  \end{columns}
\end{frame}

\section{Future challenges}

\begin{frame}{Future challenges}
  \begin{itemize}
  \item Git
  \item Cloud. Possible visions:
    \begin{itemize}
    \item As now, but `in the cloud'
    \item Integrated with `third party' compute efforts, e.g., NCI,
      NIH in the United States
    \end{itemize}
  \end{itemize}
\end{frame}

\section{Acknowledgments}

\begin{frame}{Acknowledgments}
  Core team (current \& recent): Yubo Cheng, Valerie Obenchain, Herv\'e
  Pag\`es, Marcel Ramos, Lori Shepherd, Dan Tenenbaum, Greg Wargula.
  \medskip\par
  Technical advisory board: Vincent Carey, Kasper Hansen, Wolfgang
  Huber, Robert Gentleman, Rafael Irizzary, Levi Waldron, Michael
  Lawrence, Sean Davis, Aedin Culhane
  \medskip\par
  Scientific advisory board: Simon Tavare (CRUK), Paul Flicek
  (EMBL/EBI), Simon Urbanek (AT\&T), Vincent Carey (Brigham \&
  Women's), Wolfgang Huber (EBI), Rafael Irizzary (Dana Farber),
  Robert Gentleman (23andMe)
  \medskip\par
  Research reported in this presentation was supported by the National
  Human Genome Research Institute and the National Cancer Institute of
  the National Institutes of Health under award numbers U41HG004059
  and U24CA180996. The content is solely the responsibility of the
  authors and does not necessarily represent the official views of the
  National Institutes of Health.
\end{frame}

\end{document}
